lib/rspec/mocks/proxy.rb

```diff
@@ -371,20 +371,34 @@ def original_method_handle_for(message)
 
         return super unless unbound_method
         unbound_method.bind(object)
+      rescue TypeError
+        if RUBY_VERSION == '1.8.7'
```

Comment:
If that's what you want me to do, I will do it, but frankly I think the existing code is better.

Notice that the conditional is in an exception handler, out of the main flow of the method. On post-1.8.7 interpreters, the exception handler should never be hit. Even on MRI 1.8.7, in most cases it will never be hit. So in all those cases, the performance impact should be zero. There's just this one corner case involving stubbing of singleton methods on classes, when it will actually be used.

If we extract a new method, then the performance impact will not be zero. It will have a (small) performance impact on all invocations, on all platforms.

Additionally, the code is a bit more compact this way, and there is less indirection. Indirection is a double-edged sword; it can make your code better, or it can make it worse. I feel a bit less indirection is better in this case. But it's up to you.